Solutions/Workshops/workshop4.lean

```diff
@@ -0,0 +1,331 @@
+  constructor
+  · rintro ⟨xs, xt⟩; exact ⟨xt, xs⟩
+  . rintro ⟨xt, xs⟩; exact ⟨xs, xt⟩
+  tada
+
+-- A similar statement is the theorem `Subset.antisymm`
+
+#check Subset.antisymm
+
+-- Membership to intersection and union can be unfolded using the following two theorems
+-- (although Lean usually does not need it)
+
+#check mem_union
+#check mem_inter_iff
+
+-- Exercises (MIL 4.1)
+
+-- 1
+example : s ∩ (s ∪ t) = s := by
+  ext x
+  constructor
+  . rintro ⟨xs, _⟩
+    exact xs
+  . rintro xs
+    exact ⟨xs, Or.inl xs⟩
+  tada
+-- another way:
+example : s ∩ (s ∪ t) = s := by
+  ext x
+  simp only [mem_inter_iff, mem_union, and_iff_left_iff_imp]
+  exact λ xs ↦ Or.inl xs
+
+
+-- 2
+example : s ∪ s ∩ t = s := by
+  ext x
+  constructor
+  . rintro (xs | ⟨xs, _⟩)
+    exact xs
+    exact xs
+    tada
+  . intro xs
+    exact Or.inl xs
+    tada
+  tada
+
+-- 3
+example : s \ t ∪ t = s ∪ t := by
+  ext x
+  constructor
+  . rintro (xs | xt)
+    . simp only [mem_union]
+      exact Xor'.or (Or.inl xs)
+    . right
+      exact xt
+  . rintro (xs | xt)
+    . simp only [mem_union, mem_diff]
+      by_cases h : x ∈ t
+      . right
+        exact h
+      . left
+        exact ⟨xs, h⟩
+    . right
+      exact xt
+  tada
+
+-- 4
+example : s \ t ∪ t \ s = (s ∪ t) \ (s ∩ t) := by
+  ext x
+  simp only [mem_union, mem_diff, mem_inter_iff]
+  constructor
+  . rintro (⟨xs, xnt⟩ | ⟨xs, xnt⟩) <;>
+    simp only [xnt, and_false, not_false_iff, and_true, or_false, xs, false_or]
+  .
+    rintro ⟨(xts | xts), xs⟩ <;>
+    push_neg at * <;>
+    simp only [xts, and_true, true_and, or_false, not_true,
+      and_false, false_or]
+    . exact xs xts
+    . intro h
+      exact xs h xts
+  tada
+
+-- 5
+example : s ∩ t ∪ s ∩ u ⊆ s ∩ (t ∪ u) := by
+  rintro x (⟨xs, xt⟩ | ⟨xs, xu⟩)
+  . exact ⟨xs, Or.inl xt⟩
+  . exact ⟨xs, Or.inr xu⟩
+  tada
+
+end
+
+/-
+The main difference between sets and types is the predicate `∈`. In fact, in Lean a set coincides
+with its `∈` predicate:
+
+`Set α` is defined as `α → Prop`. If `s : Set α` and `a : α`, then `a ∈ s` is defined as `s a`.
+The set `{ y | P y }` is defined as the map `fun y ↦ P y` and `h : x ∈ { y | P y }` is `h : P x`.
+If for some reason Lean fails to unfold the definition, you can use `h.out` or `rw[mem_setOf]`
+-/
+--
+
+example : 1 ∈ { x : ℕ | 0 < x } := by norm_num
+
+-- The empty set `∅` (you get this by typing `\empty`) is the function `fun y ↦ False`
+-- The universe set is the function `fun y ↦ True`
+
+example (x : ℕ) (h : x ∈ (∅ : Set ℕ)) : False :=
+  h
+
+
+example (x : ℕ) : x ∈ (univ : Set ℕ) :=
+  trivial -- `trivial` is the name of the canonical proof of `True`
+
+-- Exercises (MIL 4.1)
+
+section
+
+variable (s t : Set ℕ)
+
+def evenSet : Set ℕ :=
+  { n | Even n }
+
+def oddSet : Set ℕ :=
+  { n | ¬Even n }
+
+#print Prime
+#print Nat.Prime
+
+-- 6
+example : { n | Nat.Prime n } ∩ { n | n > 2 } ⊆ oddSet := by
+  rintro x ⟨hprime, hgt2⟩
+  suffices hodd : ¬Even x
```

Comment:
I don't think we've talked about suffices. Should we write something here as well?

Reply:
Ah, yeah probably should say something before then